octobot_trading/modes/abstract_trading_mode.py

```diff
@@ -16,32 +16,53 @@
 import abc
 import contextlib
 import decimal
+import time
+import asyncio
+import concurrent.futures
 
+import octobot_commons.channels_name as channels_name
 import octobot_commons.constants as common_constants
+import octobot_commons.enums as common_enums
 import octobot_commons.logging as logging
+import octobot_commons.databases as databases
+import octobot_commons.tree as commons_tree
+import octobot_commons.configuration as commons_configuration
 import octobot_commons.tentacles_management as abstract_tentacle
 import octobot_commons.authentication as authentication
 
 import async_channel.constants as channel_constants
+import async_channel.channels as channels
 
 import octobot_tentacles_manager.api as tentacles_manager_api
 import octobot_tentacles_manager.configuration as tm_configuration
 
+import octobot_backtesting.api as backtesting_api
+
+import octobot_trading.util as util
 import octobot_trading.constants as constants
 import octobot_trading.enums as enums
 import octobot_trading.exchange_channel as exchanges_channel
 import octobot_trading.modes.script_keywords as script_keywords
+import octobot_trading.modes.modes_factory as modes_factory
+import octobot_trading.modes.channel.abstract_mode_producer as abstract_mode_producer
+import octobot_trading.modes.channel.abstract_mode_consumer as abstract_mode_consumer
 import octobot_trading.personal_data.orders as orders
+import octobot_trading.personal_data.portfolios as portfolios
+import octobot_trading.personal_data.trades as personal_data_trades
 import octobot_trading.signals as signals
+import octobot_trading.modes.script_keywords.basic_keywords as basic_keywords
+import octobot_trading.modes.script_keywords.context_management as context_management
 
 
 class AbstractTradingMode(abstract_tentacle.AbstractTentacle):
     __metaclass__ = abc.ABCMeta
+    USER_INPUT_TENTACLE_TYPE = common_enums.UserInputTentacleTypes.TRADING_MODE
 
     MODE_PRODUCER_CLASSES = []
     MODE_CONSUMER_CLASSES = []
     # maximum seconds before sending a trading signal if orders are slow to create on exchange
     TRADING_SIGNAL_TIMEOUT = 10
+    SAVED_RUN_METADATA_DB_BY_BOT_ID = {}
```

Comment:
I don't understand why we need this static attribute here (instead of having a class attribute)

Reply:
We need to have a way to know if we already saved the run metadata or not for a given bot. We don't want to save the same metadata as many times as there are instances of trading modes (which is one per symbol per exchange, usually)